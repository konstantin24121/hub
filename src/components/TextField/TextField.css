.root {
  $root: &;

  position: relative;
  height: 48px;
  line-height: 24px;
  font-size: 1rem;

  &_isFocused {
    /* Focus state */
  }

  &_isDirty {
    /* When user touch field */
  }

  &__placeholder {
    position: absolute;
    bottom: 12px;
    opacity: 0;
    color: rgba(0, 0, 0, 0.298039);
    transition: opacity 350ms cubic-bezier(0.23, 1, 0.32, 1);

    &_isVisible {
      opacity: 1;
    }
  }

  &__input {
    /* Clean default styles */
    border: none;
    outline: none;
    height: 100%;
    width: 100%;
    background-color: transparent;
  }

  &__underline_static,
  &__underline_dynamic {
    box-sizing: content-box;
    border-width: 0;
    border-color: rgb(224, 244, 244);
    border-style: solid;
    border-bottom-width: 1px;
    margin: 0;
    position: absolute;
    width: 100%;
    height: 0;
    bottom: 8px;
  }

  &__underline_dynamic {
    border-color: rgb(0, 188, 212);
    border-bottom-width: 2px;
    transform: scaleX(0);
    transition: transform 350ms cubic-bezier(0.23, 1, 0.32, 1);

    ^&_isFocused &,
    ^&:focus & {
      transform: scaleX(1);
    }
  }
}
