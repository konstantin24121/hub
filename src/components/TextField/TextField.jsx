import React, { PureComponent, PropTypes } from 'react';
// Helpers
import cn from 'classnames';
import capitalize from 'lodash/capitalize';

import s from './TextField.css';

/**
 * Поле для ввода текста, компонент может использоватся
 * как обычный input, так и примитивный textarea
 */
class TextField extends PureComponent {
  static propTypes = {
    /**
     * Имя поля
     */
    name: PropTypes.string.isRequired,
    /**
     * Значение поля
     */
    value: PropTypes.string,
    /**
     * Placeholder для поля
     */
    placeholder: PropTypes.string,
    /**
     * Плавающий label
     */
    floatingLabel: PropTypes.string,
    /**
     * Подсказка для поля ввода
     * самое распространненное использование в качестве
     * показа ошибок валидации
     */
    hint: PropTypes.string,
    /**
     * Статус поля
     */
    status: PropTypes.oneOf(['normal', 'warning', 'danger']),
    /**
     * Срабатывает при изменении value в поле ввода
     */
    onChange: PropTypes.func,
    /**
     * Срабатывает при получении фокуса полем
     */
    onFocus: PropTypes.func,
    /**
     * Срабатывает при потере фокуса полем
     */
    onBlur: PropTypes.func,
  };

  static defaultProps = {
    value: '',
    placeholder: '',
    floatingLabel: '',
    status: 'normal',
    /* eslint-disable no-unused-vars */
    onChange: ({ value }, e) => {},
    onFocus: (event) => {},
    onBlur: (event) => {},
    /* eslint-enable no-unused-vars */
  };

  constructor(props) {
    super(props);
    this.state = {
      value: props.value,
      isFocused: false,
      /**
       * Поле было затронуто пользователем
       * как церковный мальчик католическим священником
       */
      isDirty: false,
    };
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      value: nextProps.value,
    });
  }

  /**
   * Handles
   */
  handleFocus = (e) => {
    this.setState({
      isFocused: true,
      isDirty: true,
    });
    this.props.onFocus(e);
  };

  handleBlur = (e) => {
    this.setState({
      isFocused: false,
    });
    this.props.onBlur(e);
  };

  handleChange = (e) => {
    const { value } = e.target;
    this.setState({ value });
    this.props.onChange({ value }, e);
  }

  /**
   * Renders
   */
  render() {
    const { name, placeholder, floatingLabel, status, hint } = this.props;
    const { value, isFocused, isDirty } = this.state;

    const isEmpty = !value;
    const hasFloatingLabel = !!floatingLabel;
    const isLabelFloating = !isEmpty || isFocused;

    const rootStyle = cn(s.root, s[`root_is${capitalize(status)}`], {
      [s.root_isFocused]: isFocused,
      [s.root_isDirty]: isDirty,
      [s.root_hasFloatingLabel]: hasFloatingLabel,
      [s.root_hasHint]: !!hint,
    });
    const labelStyle = cn(s.root__label, {
      [s.root__label_isFloat]: isLabelFloating,
    });
    const placeholderStyle = cn(s.root__placeholder, {
      [s.root__placeholder_isVisible]: (isLabelFloating && isEmpty)
      || (!hasFloatingLabel && isEmpty),
    });
    const inputStyle = cn(s.root__input);
    const underlineStyle = cn(s.root__underline);
    const underlineStaticStyle = cn(s.root__underline_static);
    const underlineDynamicStyle = cn(s.root__underline_dynamic);
    const hintStyle = cn(s.root__hint);

    return (
      <div className={rootStyle}>
        {placeholder && <div className={placeholderStyle}>{placeholder}</div>}
        {floatingLabel && <div className={labelStyle}>{floatingLabel}</div>}
        <input
          type="text"
          name={name}
          value={value}
          className={inputStyle}
          onChange={this.handleChange}
          onFocus={this.handleFocus}
          onBlur={this.handleBlur}
        />
        <div className={underlineStyle}>
          <hr className={underlineStaticStyle} />
          <hr className={underlineDynamicStyle} />
        </div>
        {hint && <div className={hintStyle}>{hint}</div>}
      </div>
    );
  }
}

export default TextField;
/**
 * version: 0.0.1
 */
