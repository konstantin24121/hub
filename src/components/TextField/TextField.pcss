.root {
  display: inline-block;
  width: 100%;
  position: relative;
  height: 48px;
  line-height: 24px;
  font-size: 1rem;

  &__is_normal {
    /* Normal state */
    color: rgb(0, 188, 212);
  }

  &__is_warning {
    color: rgb(255, 152, 0);
  }

  &__is_danger {
    color: rgb(244, 67, 54);
  }

  &__is_focused {
    /* Focus state */
  }

  &__is_dirty {
    /* When user touch field */
  }

  &__is_disabled {
    opacity: 0.8;
    user-select: none;
    pointer-events: none;
    cursor: not-allowed;
  }

  &__has_floating_label {
    margin-top: 14px;
  }

  &__has_hint {
    margin-bottom: 14px;
  }
}

.placeholder {
  position: absolute;
  bottom: 12px;
  opacity: 0;
  color: rgba(0, 0, 0, 0.298039);
  transition: opacity 350ms cubic-bezier(0.23, 1, 0.32, 1);
  user-select: none;
  pointer-events: none;
  z-index: 1;

  &__is_visible {
    opacity: 1;
  }
}

.label {
  position: absolute;
  bottom: 12px;
  color: rgba(0, 0, 0, 0.298039);
  transform-origin: left top 0;
  transition-property: transform, color;
  transition-duration: 450ms;
  transition-timing-function: cubic-bezier(0.23, 1, 0.32, 1);
  user-select: none;
  pointer-events: none;
  z-index: 1;

  &__is_float {
    color: inherit;
    transform: scale(0.75) translate(0, -28px);
  }
}

.input {
  /* Clean default styles */
  border: none;
  outline: none;
  height: 100%;
  width: 100%;
  background-color: transparent;
  position: relative;
  z-index: 2;
  cursor: inherit;
}

.underline_static,
.underline_dynamic {
  box-sizing: content-box;
  border-width: 0;
  border-color: rgb(224, 244, 244);
  border-style: solid;
  border-bottom-width: 1px;
  margin: 0;
  position: absolute;
  width: 100%;
  height: 0;
  bottom: 8px;

  .root__is_danger & {
    border-color: rgb(244, 67, 54);

    /* Progressive */
    border-color: currentColor;
  }
}

.underline_dynamic {
  border-bottom-width: 2px;
  transform: scaleX(0);
  transition: transform 350ms cubic-bezier(0.23, 1, 0.32, 1);

  .root__is_normal & {
    border-color: rgb(0, 188, 212);

    /* Progressive */
    border-color: currentColor;
  }

  .root__is_warning & {
    border-color: rgb(255, 152, 0);

    /* Progressive */
    border-color: currentColor;
  }

  .root__is_danger & {
    border-color: rgb(244, 67, 54);

    /* Progressive */
    border-color: currentColor;
  }

  .root__is_focused &,
  .root:focus & {
    transform: scaleX(1);
  }
}

.hint {
  position: relative;
  font-size: 0.75em;
  color: inherit;
  margin-top: -9px;
}
