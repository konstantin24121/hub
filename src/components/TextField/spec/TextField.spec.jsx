import React from 'react';
import TextField from '../index';

describe('TextField', () => {
  it('Render a TextField with correct name', () => {
    const wrapper = shallow(
      <TextField name="field" />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
