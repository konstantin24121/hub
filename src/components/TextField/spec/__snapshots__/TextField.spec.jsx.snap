// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Focus at TextField with float label mast have effect 1`] = `
<div
  className="root root_isNormal root_hasFloatingLabel"
>
  <div
    className="label"
  >
    label
  </div>
  <input
    className="input"
    disabled={false}
    name="field"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <div
    className="underline"
  >
    <hr
      className="underlineStatic"
    />
    <hr
      className="underlineDynamic"
    />
  </div>
</div>
`;

exports[`TextField Focus at TextField with float label mast have effect 2`] = `
<div
  className="root root_isNormal root_isFocused root_isDirty root_hasFloatingLabel"
>
  <div
    className="label label_isFloat"
  >
    label
  </div>
  <input
    className="input"
    disabled={false}
    name="field"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <div
    className="underline"
  >
    <hr
      className="underlineStatic"
    />
    <hr
      className="underlineDynamic"
    />
  </div>
</div>
`;

exports[`TextField Render a TextField with correct props and default callbacks are fired 1`] = `
<div
  className="root root_isNormal root_hasHint"
>
  <div
    className="placeholder placeholder_isVisible"
  >
    Placehodor
  </div>
  <input
    className="input"
    disabled={false}
    name="field"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <div
    className="underline"
  >
    <hr
      className="underlineStatic"
    />
    <hr
      className="underlineDynamic"
    />
  </div>
  <div
    className="hint"
  >
    Hint Ladger
  </div>
</div>
`;

exports[`TextField Render a TextField with correct props and default callbacks are fired 2`] = `
<div
  className="root root_isNormal root_isFocused root_isDirty root_hasHint"
>
  <div
    className="placeholder placeholder_isVisible"
  >
    Placehodor
  </div>
  <input
    className="input"
    disabled={false}
    name="field"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <div
    className="underline"
  >
    <hr
      className="underlineStatic"
    />
    <hr
      className="underlineDynamic"
    />
  </div>
  <div
    className="hint"
  >
    Hint Ladger
  </div>
</div>
`;

exports[`TextField Render a TextField with correct props and default callbacks are fired 3`] = `
<div
  className="root root_isNormal root_isDirty root_hasHint"
>
  <div
    className="placeholder placeholder_isVisible"
  >
    Placehodor
  </div>
  <input
    className="input"
    disabled={false}
    name="field"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value=""
  />
  <div
    className="underline"
  >
    <hr
      className="underlineStatic"
    />
    <hr
      className="underlineDynamic"
    />
  </div>
  <div
    className="hint"
  >
    Hint Ladger
  </div>
</div>
`;

exports[`TextField Render a TextField with correct props and default callbacks are fired 4`] = `
<div
  className="root root_isNormal root_isDirty root_hasHint"
>
  <div
    className="placeholder"
  >
    Placehodor
  </div>
  <input
    className="input"
    disabled={false}
    name="field"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="nezad"
  />
  <div
    className="underline"
  >
    <hr
      className="underlineStatic"
    />
    <hr
      className="underlineDynamic"
    />
  </div>
  <div
    className="hint"
  >
    Hint Ladger
  </div>
</div>
`;
